src/sentry/notifications/helpers.py

```diff
@@ -833,9 +833,141 @@ def get_setting_options_for_recipient(
+                user_to_setting[ns.user][setting_type] = NotificationSettingsOptionEnum(ns.value)
+
+    defaults = get_type_defaults()
+    for _, ns_dict in user_to_setting.items():
+        for type in NotificationSettingEnum:
+            if type not in ns_dict and type in defaults:
+                user_to_setting[ns.user][type] = defaults[type]
+
+    return user_to_setting
+
+
+def get_setting_providers_for_users(
+    user_ids: Iterable[int],
+    project: Project | None = None,
+    organization: Organization | None = None,
+) -> MutableMapping[RpcActor, MutableMapping[ExternalProviderEnum, NotificationSettingProvider]]:
+    """
+    Returns a map of users to NotificationSettingProvider by provider:
+    {user -> {provider -> setting}}.
+
+    Args:
+        user_ids: The user ids to get notification settings for.
+        project: The project to get notification settings for.
+        organization: The organization to get notification settings for.
+    """
+    query = get_query(project=project, organization=organization, user_ids=user_ids)
+    notification_settings = NotificationSettingProvider.objects.filter(query)
+
+    user_to_setting: MutableMapping[
+        RpcActor, MutableMapping[ExternalProviderEnum, NotificationSettingProvider]
+    ] = defaultdict(dict)
+    for ns in notification_settings:
+        if ns.scope_type == NotificationScopeEnum.PROJECT.value:
+            provider = ExternalProviderEnum(ns.provider)
+            user_to_setting[ns.user][provider] = ns
+
+    for ns in notification_settings:
+        if ns.scope_type == NotificationScopeEnum.ORGANIZATION.value:
+            ns_dict = user_to_setting[ns.user]
+            provider = ExternalProviderEnum(ns.provider)
+            if provider not in ns_dict:
+                ns_dict[provider] = ns
+
+    for ns in notification_settings:
+        if ns.scope_type == NotificationScopeEnum.USER.value:
+            ns_dict = user_to_setting[ns.user]
+            provider = ExternalProviderEnum(ns.provider)
+            if provider not in ns_dict:
+                ns_dict[provider] = ns
+
+    # TODO: Are there any provider defaults to fill in?
+    return user_to_setting
+
+
+def get_notification_recipients(project: Project) -> Mapping[ExternalProviderEnum, set[RpcActor]]:
+    """
+    Returns the recipients that should be notified for each provider.
+
+    Args:
+        project: The project to get notification settings for.
+    """
+    user_ids = project.member_set.values_list("user_id", flat=True)
+    options = get_setting_options_for_users(user_ids, project=project)
+    providers = get_setting_providers_for_users(user_ids, project=project)
+
+    enabled_providers: Mapping[
+        RpcActor, Mapping[ExternalProviderEnum, NotificationSettingProvider]
+    ] = {
+        user: {
+            provider: ps
+            for provider, ps in setting.items()
+            if ps.value != NotificationSettingsOptionEnum.NEVER.value
```

Comment:
@snigdhas what about default values here? Do we need to account for them?

Reply:
The defaults should be included in the return value for `get_setting_providers_for_users`.